mongo/writeconcern/writeconcern.go

```diff
@@ -17,34 +21,126 @@ import (
 var ErrNegativeW = errors.New("write concern `w` field cannot be a negative number")
 
 // ErrNegativeWTimeout indicates that a negative WTimeout was specified.
+//
+// Deprecated: ErrNegativeWTimeout will be removed in Go Driver 2.0.
 var ErrNegativeWTimeout = errors.New("write concern `wtimeout` field cannot be negative")
 
-// WriteConcern describes the level of acknowledgement requested from MongoDB for write operations
-// to a standalone mongod or to replica sets or to sharded clusters.
+// A WriteConcern defines a MongoDB read concern, which describes the level of acknowledgment
+// requested from MongoDB for write operations to a standalone mongod, to replica sets, or to
+// sharded clusters.
+//
+// For more information about MongoDB write concerns, see
+// https://www.mongodb.com/docs/manual/reference/write-concern/
 type WriteConcern struct {
-	w interface{}
-	j bool
+	// W requests acknowledgment that the write operation has propagated to a specified number of
+	// mongod instances or to mongod instances with specified tags. It sets the the "w" option
+	// in a MongoDB write concern.
+	//
+	// W must be a type string or int value.
+	//
+	// For more information about the "w" option, see
+	// https://www.mongodb.com/docs/manual/reference/write-concern/#w-option
+	W interface{}
+
+	// Journal requests acknowledgment from MongoDB that the write operation has been written to the
+	// on-disk journal. It sets the "j" option in a MongoDB write concern.
+	//
+	// For more information about the "j" option, see
+	// https://www.mongodb.com/docs/manual/reference/write-concern/#j-option
+	Journal *bool
+
+	// WTimeout specifies a time limit for the write concern. It sets the "wtimeout" option in a
+	// MongoDB write concern.
+	//
+	// It is only applicable for "w" values greater than 1. Using a WTimeout and setting Timeout on
+	// the Client at the same time will result in undefined behavior.
+	//
+	// For more information about the "wtimeout" option, see
+	// https://www.mongodb.com/docs/manual/reference/write-concern/#wtimeout
+	WTimeout time.Duration
+}
 
-	// NOTE(benjirewis): wTimeout will be deprecated in a future release. The more general Timeout
-	// option may be used in its place to control the amount of time that a single operation can run
-	// before returning an error. Using wTimeout and setting Timeout on the client will result in
-	// undefined behavior.
-	wTimeout time.Duration
+// Majority returns a WriteConcern that requests acknowledgment that write operations have been
+// durably committed to the calculated majority of the data-bearing voting members. If journaling is
+// enabled on mongod, nodes only acknowledge when the write operation has been written to the
+// on-disk journal.
+//
+// For more information about write concern "w: majority", see
+// https://www.mongodb.com/docs/manual/reference/write-concern/#mongodb-writeconcern-writeconcern.-majority-
+func Majority() *WriteConcern {
```

Comment:
The "convenience" API as currently implemented in this PR is:
```go
writeconcern.W0()
writeconcern.W1()
writeconcern.Majority()
```

A reasonable alternative could be:
```go
writeconcern.Unacknowledged()
writeconcern.Acknowledged()
writeconcern.Journaled()
writeconcern.Majority()
```

Ignoring name collisions, a reasonable alternative could be:
```go
writeconcern.W(0)
writeconcern.W(1)
writeconcern.WMajority()
```

We could also consider not providing any convenience functions at all:
```go
func ptr[T any](v T) *T { return &v }
&writeconcern.WriteConcern{W: 0}
&writeconcern.WriteConcern{W: 1}
&writeconcern.WriteConcern{W: "majority"}
```

If the

Reply:
(1) and (2) are easy to understand and follow the documentation conventions. It seems reasonable to me to keep the existing API. Perhaps once W() is gone we can also include a W() constructor if there is demand.

Reply:
I decided to mix (1) and (2) and ended up with:
```go
writeconcern.Unacknowledged()
writeconcern.W1()
writeconcern.Journaled()
writeconcern.Majority()
```
That gives us the API "space" to add additional convenience functions if we need to in the future.

Reply:
Looking at Java's `WriteConcern` class, [ACKNOWLEDGED](https://github.com/mongodb/mongo-java-driver/blob/0efc78c41c63e91da23532695fc31ee1dbd157d0/driver-core/src/main/com/mongodb/WriteConcern.java#L95) is actually the empty (i.e. default) write concern, and there is a separate [W1](https://github.com/mongodb/mongo-java-driver/blob/0efc78c41c63e91da23532695fc31ee1dbd157d0/driver-core/src/main/com/mongodb/WriteConcern.java#L103). While it is correct to say that the default write concern and write concern `{w: 1}` are acknowledged, I was worried using `Acknowledged` for a different acknowledged write concern than the Java driver might be confusing (`Acknowledged` is a bit confusing in general because most write concerns are acknowledged). Instead, I decided to start with the most obvious write concerns that are also consistent with the Java driver naming. That leaves the option open to define more write concerns later if we determine it's necessary.

Reply:
Thanks for the explanation